File: taverna-scufl2-examples/src/main/java/org/apache/taverna/examples/ProcessorNames.java
package org.apache.taverna.examples;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */


import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreeModel;
import javax.xml.bind.JAXBException;

import org.apache.taverna.scufl2.api.common.Scufl2Tools;
import org.apache.taverna.scufl2.api.common.URITools;
import org.apache.taverna.scufl2.api.container.WorkflowBundle;
import org.apache.taverna.scufl2.api.core.Processor;
import org.apache.taverna.scufl2.api.core.Workflow;
import org.apache.taverna.scufl2.api.io.ReaderException;
import org.apache.taverna.scufl2.api.io.WorkflowBundleIO;
import org.apache.taverna.scufl2.api.profiles.Profile;

public class ProcessorNames {

	public static void main(String[] args) throws JAXBException, IOException,
			ReaderException {
		WorkflowBundleIO io = new WorkflowBundleIO();
		ProcessorNames processorNames = new ProcessorNames();
		for (String filename : args) {
			// NOTE: We give 'null' as filetype so it will guess based on filename
			WorkflowBundle ro = io.readBundle(new File(filename), null);
//			System.out.print(filename + ": ");
//			System.out.println(processorNames.showProcessorNames(ro));
			System.out.println(processorNames.showProcessorTree(ro));
		}
	}
	private Scufl2Tools scufl2Tools = new Scufl2Tools();

	private URITools uriTools = new URITools();

	private Workflow findNestedWorkflow(Processor processor) {
		Profile profile = processor.getParent().getParent().getMainProfile();
		return scufl2Tools.nestedWorkflowForProcessor(processor, profile);
	}

	private void findProcessors(WorkflowBundle ro, Workflow workflow,
			DefaultMutableTreeNode parent) {
		for (Processor processor : workflow.getProcessors()) {
			DefaultMutableTreeNode processorNode = new DefaultMutableTreeNode(
					processor.getName());
			parent.add(processorNode);
			Workflow wf = findNestedWorkflow(processor);
			if (wf != null) {
				findProcessors(ro, wf, processorNode);
			}
		}

	}

	public TreeModel makeProcessorTree(WorkflowBundle workflowBundle)
			throws JAXBException, IOException {
		Workflow workflow = workflowBundle.getMainWorkflow();
		TreeModel treeModel = new DefaultTreeModel(new DefaultMutableTreeNode(
				workflow.getName()));
		DefaultMutableTreeNode parent = (DefaultMutableTreeNode) treeModel
				.getRoot();

		findProcessors(workflowBundle, workflow, parent);
		return treeModel;
	}

	public List<String> showProcessorNames(WorkflowBundle ro)
			throws JAXBException, IOException {
		ArrayList<String> names = new ArrayList<String>();
		for (Processor processor : ro.getMainWorkflow().getProcessors()) {
			names.add(processor.getName());
		}
		Collections.sort(names);
		return names;
	}

	public String showProcessorTree(WorkflowBundle ro) throws JAXBException,
			IOException {
		TreeModel treeModel = makeProcessorTree(ro);
		return treeModelAsString(treeModel);
	}

	public String treeModelAsString(TreeModel treeModel) {
		StringBuffer sb = new StringBuffer();
		Object root = treeModel.getRoot();
		treeModelAsString(treeModel, root, sb, "");
		return sb.toString();
	}

	protected void treeModelAsString(TreeModel treeModel, Object parent,
			StringBuffer sb, String indentation) {
		sb.append(indentation);
		int childCount = treeModel.getChildCount(parent);
		if (childCount == 0) {
			sb.append("- ");
		} else {
			sb.append("+ ");
			indentation = indentation + "  ";
		}
		sb.append(parent);
		sb.append("\n");
		for (int i = 0; i < childCount; i++) {
			Object child = treeModel.getChild(parent, i);
			treeModelAsString(treeModel, child, sb, indentation);
		}
	}
}

File: taverna-scufl2-examples/src/main/python/processorNames.py
#!/usr/bin/env python
#
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
#

import sys
import zipfile
import os.path
import xml.etree.ElementTree
from rdflib import Graph
from rdflib import Namespace
import urllib
from rdflib.term import URIRef

Scufl2NS = Namespace("http://ns.taverna.org.uk/2010/scufl2#")


NS_CONTAINER="{urn:oasis:names:tc:opendocument:xmlns:container}"


import rdflib

class Scufl2(object):
    def __init__(self, filename=None):
        self.filename = filename
        if filename is not None:
            self.zip = zipfile.ZipFile(filename)
            self.check_mime_type()
            self.parse()    

    def check_mime_type(self):       
        mimetype = self.zip.open("mimetype").read()
        if mimetype != "application/vnd.taverna.scufl2.workflow-bundle":
            raise Scufl2Error("Unknown mimetype %r" % mimetype)

    def _is_valid_prefix(self, filename):
        valid_prefixes = ["workflow/", "annotation/", "profile/"]
        for prefix in valid_prefixes:
            if filename.startswith(prefix):
                return True
        return False        

    def parse_all_graphs(self, sameAs):
        filenames = {}
        for filename in self.zip.namelist():
            if not self._is_valid_prefix(filename):
                continue
            if filename.endswith(".ttl"):
                base = filename[:-4]
                filenames[base] = filename
            elif filename.endswith(".rdf"):
                base = filename[:-4]
                if not base in filenames:
                    filenames[base] = filename
        
        for name in filenames:
            filename = filenames[name]
            rdf_file = self.zip.open(filename)
            format = "n3"
            if filename.endswith(".rdf"):
                format = "xml"
            base = sameAs + filename
            self.graph.parse(rdf_file, base, format=format)

    def parse(self):
        if "workflowBundle.ttl" in self.zip.namelist():
            format = "n3" 
            rootfile = "workflowBundle.ttl"
        elif "workflowBundle.rdf" in self.zip.namelist():
            rootfile = "workflowBundle.rdf"
            format = "xml" 
        else:
            raise Scufl2Error("Can't find workflowBundle.ttl or "
                              "workflowBundle.rdf")

        self.uri = "file://" + urllib.pathname2url(os.path.abspath(self.filename)) + "/"
        early_graph = Graph()    
        rdf_file = self.zip.open(rootfile)
        early_graph.parse(rdf_file, self.uri, format=format)
        sameBaseAs = list(early_graph.objects(subject=URIRef(self.uri), predicate=Scufl2NS.sameBaseAs))

        if not sameBaseAs:
            # Fall back to the file:/// URIs   
            self.graph = early_graph
        else:    
            # Use the sameBaseAs as the base
            self.uri = sameBaseAs[0]
            self.graph = Graph()
            # Reparse it
            rdf_file = self.zip.open(rootfile)
            self.graph.parse(rdf_file, self.uri, format=format)

        self.parse_all_graphs(self.uri)


class Scufl2Error(Exception):
    pass


def main(prg="processorNames.py", filename=None):
    if filename is None:
        filename = os.path.join(os.path.dirname(prg), "..", "..", "..", "..", 
        "scufl2-rdfxml", "src", "test", "resources",
        "uk", "org", "taverna","scufl2","rdfxml", "example.wfbundle")
    
    scufl2 = Scufl2(filename)
    
    for workflowUri in scufl2.graph.objects(predicate=Scufl2NS.workflow):
        for name in scufl2.graph.objects(workflowUri, Scufl2NS.name):
            print name
        for processorUri in scufl2.graph.objects(workflowUri, Scufl2NS.processor):
            for name in scufl2.graph.objects(processorUri, Scufl2NS.name):
                print "---", name, processorUri


if __name__ == "__main__":
    main(*sys.argv)

